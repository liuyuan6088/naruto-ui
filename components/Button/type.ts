import { tuple } from '../utils/type';

export const ButtonTypes = tuple('default', 'primary', 'dashed', 'danger');
export type ButtonType = (typeof ButtonTypes)[number];

export const ButtonSizes = tuple('large', 'default', 'small');
export type ButtonSize = (typeof ButtonSizes)[number];

export const ButtonHTMLTypes = tuple('submit', 'button', 'reset');
export type ButtonHTMLType = (typeof ButtonHTMLTypes)[number];

export type AnchorButtonProps = {
  href: string;
  target?: string;
  onClick?: React.MouseEventHandler<HTMLAnchorElement>;
} & IButtonProps &
  React.AnchorHTMLAttributes<HTMLAnchorElement>;

export interface IButtonProps {
  prefixCls?: string,
  disabled?: boolean,
  type?: ButtonType,
  size?: ButtonSize,
  block?: boolean,
  htmlType?: ButtonHTMLType,
  className?: string,
  style?: React.CSSProperties
  href?: string,
  onClick?: (e: React.MouseEvent) => void,
}

export interface ButtonGroupProps {
  size?: ButtonSize;
  style?: React.CSSProperties;
  className?: string;
  prefixCls?: string;
}
