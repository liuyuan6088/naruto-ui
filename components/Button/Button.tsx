import * as React from 'react'
import cx from 'classnames'
import omit from 'omit.js'
import Group from './ButtonGroup'
import Icon from '../Icon'
import { primaryName } from '../utils/constant'
import { IButtonProps, AnchorButtonProps } from './type'
import './style/index.less'

const defaultProps: IButtonProps = {
  prefixCls: `${primaryName}-btn`,
  type: 'default',
  block: false
}

const handleClick = ({ onClick, disabled, loading }: IButtonProps) => (e: React.MouseEvent) => {
  if (disabled) {
    return
  }
  if (loading) return
  if (onClick) {
    onClick(e)
  }
}

const renderChildren = (children: React.ReactNode) => {
  if (!(children || children === 0)) {
    return null
  }
  if (typeof children === 'string') {
    return <span>{children}</span>
  }
  return children
}

// 解决写了默认值 使用仍然需要必填的问题
const Button: React.FC<IButtonProps> & {
  defaultProps: Partial<IButtonProps>
  Group?: typeof Group
} = props => {
  const {
    prefixCls,
    disabled,
    loading,
    type,
    size,
    block,
    htmlType,
    className,
    style,
    href,
    icon,
    children,
    ...rest
  } = props

  const classes = cx(prefixCls, className, {
    [`${prefixCls}-block`]: block,
    [`${prefixCls}-loading`]: loading,
    [`${prefixCls}-${type}`]: type,
    [`${prefixCls}-${size}`]: size
  })

  const iconType = loading ? 'reload' : icon
  const iconNode = iconType ? <Icon type={iconType} spin className={`${prefixCls}-icon`} /> : null

  if (href) {
    const linkRestProps = omit(rest as AnchorButtonProps, ['htmlType'])
    const classesA = cx(`${prefixCls}-a`, className, {
      [`${prefixCls}-block`]: block,
      [`${prefixCls}-${size}`]: size
    })
    return (
      <a
        {...linkRestProps}
        href={href}
        style={style}
        className={classesA}
        onClick={handleClick(props)}
      >
        {renderChildren(children)}
      </a>
    )
  }

  return (
    <button
      type={htmlType || 'button'}
      style={style}
      className={classes}
      disabled={disabled}
      onClick={handleClick(props)}
    >
      {iconNode}
      {renderChildren(children)}
    </button>
  )
}

Button.defaultProps = defaultProps

export default Button
