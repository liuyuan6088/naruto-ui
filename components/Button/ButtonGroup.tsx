import * as React from 'react'
import cx from 'classnames'
import { primaryName } from '../utils/constant'
import { ButtonGroupProps } from './type'
import './style/index.less'

const ButtonGroup: React.FC<ButtonGroupProps> = props => {
  const { prefixCls, size, className, ...rest } = props

  const classes = cx(prefixCls, className, {
    [`${prefixCls}-${size}`]: size
  })

  return <div {...rest} className={classes} />
}

ButtonGroup.defaultProps = {
  prefixCls: `${primaryName}-btn-group`
}

export default ButtonGroup
