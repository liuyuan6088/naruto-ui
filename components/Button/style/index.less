@import '../../styles/theme/default.less';

@btn-prefix-cls: ~"@{primary-name}-btn";

.@{btn-prefix-cls} {
  outline: none;
  cursor: pointer;
  white-space: nowrap;
  user-select: none;
  text-align: center;
  color: @gray-0;
  font-size: @font-size-small;
  background-color: transparent;
  border: 1px solid @border-color;
  padding: 0 @btn-space;
  border-radius: @bordor-radius;
  transition: all .3s @ease-in-out; 
  height: 32px;
  line-height: 1;

  &-a {
    display: inline-block;
    color: @gray-0;
    height: 32px;
    padding: 0 @btn-space;
    transition: all .3s @ease-in-out;
    font-size: @font-size-small;
    > span {
      display: inline-block;
      height: 100%;
    }
    &:hover {
      color: @primary-color;
    }
  }

  &-default {
    background-color: #fff;
    &:hover {
      color: @primary-color;
      border-color: @primary-color;
    }
  }

  &-primary {
    background-color: @primary-color;
    color: @white;
    border-color: transparent;
    &:hover {
      opacity: 0.8;
    }
  }

  &-dashed {
    border-style: dashed;
    &:hover {
      color: @primary-color;
      border-color: @primary-color;
    }
  }

  &-danger {
    color: @red-0;
    background-color: #f5f5f5;
    &:hover {
      color: @white;
      background-color: @red-0;
      border-color: transparent;
    }
  }

  &-small {
    .small();
    padding: 0 @btn-space-sm;
  }

  &-large {
    .large();
    font-size: @font-size-large;
  }

  &-block {
    width: 100%;
  }

  &.disabled,
  &[disabled] {
    color: rgba(0,0,0,0.25);
    background-color: @disabled;
    border-color: #d9d9d9;
    cursor: not-allowed;
    > * {
      pointer-events: none;
    }
    &:hover {
      opacity: 1;
    }
  }

  &-link {
    border: none;
    color: @primary-color;
    &:hover {
      opacity: .8;
    }
    &[disabled] {
      background-color: transparent;
    }
  }

  &-loading {
    opacity: .8;
  }

  &-icon {
    margin-right: 5px;
    vertical-align: text-top;
  }

  // group
  .@{btn-prefix-cls}-group &:not(:first-child):not(:last-child) {
    // border-right-color: @primary-color;
    // border-left-color: @primary-color;
  }
  &-group {
    .@{btn-prefix-cls} {
      border-radius: 0;
      &:not(:first-child) {
        margin-left: -1px;
      }
      &:hover {
        position: relative; // 提升层级
        z-index: 0;
        border-right-color: @primary-color;
      }
      &:focus {
        position: relative; // 提升层级
        z-index: 1;
        border-color: @primary-color;
      }
      &:first-child {
        border-bottom-left-radius: @bordor-radius;
        border-top-left-radius: @bordor-radius;
      }
      &:last-child {
        border-bottom-right-radius: @bordor-radius;
        border-top-right-radius: @bordor-radius;
      }
      &-primary {
        position: relative;
        &:not(:last-child) {
          &::after {
            content: '';
            position: absolute;
            height: 100%;
            width: 1px;
            background-color: #fff;
            top: 0;
            right: 0;
          }
        }
        &:focus {
          opacity: .8;
        }
      }
    }
  }
  &-group-small {
    .@{btn-prefix-cls} {
      .small();
    }
  }

  &-group-large {
    .@{btn-prefix-cls} {
      .large();
    }
  }
}

.small {
  height: 24px;
}

.large {
  height: 40px;
}
