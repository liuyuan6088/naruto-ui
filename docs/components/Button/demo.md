---
imports:
  import ButtomDemo from './demo.tsx';
---

# Button 按钮

基础组件，触发业务逻辑时使用。

### 简单示例

:::demo
  <ButtomDemo>dads</ButtomDemo>
:::

## Props
| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
