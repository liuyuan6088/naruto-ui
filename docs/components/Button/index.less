
.button {
  &-box {
    button + button {
      margin-left: 20px;
    }
  }
}

.naruto-btn-group {
  + .naruto-btn-group {
    margin-top: 20px;
  }
}