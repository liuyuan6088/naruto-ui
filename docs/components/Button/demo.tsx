import * as React from 'react';
import { Button } from 'components';

const ButtonGroup = Button.Group;

const ButtonDemo: React.FC = () => {

  const handleClick = (e: React.MouseEvent) => {
    console.log(e);
  }

  return (
    <React.Fragment>
      <Button>Default</Button>
      <Button type='primary'>Primary</Button>
      <Button type='dashed'>Dashed</Button>
      <Button type='danger'>Danger</Button>

      <Button type='primary' size='large'>Primary</Button>
      <Button type='primary' size='small'>Primary</Button>

      <Button type='primary' disabled={true}>Primary</Button>
      <Button type='primary' block={true}>Primary</Button>

      <Button type='primary' onClick={handleClick}>Primary</Button>

      <Button href='http://www.baidu.com'>aaa</Button>

      <Button type='primary' loading>Primary</Button>

      <ButtonGroup>
        {['left', 'mid', 'right'].map((e) => <Button
          key={e}
          type='primary'
          onClick={data => {
            console.log(data);
          }}
        >{e}</Button>)}
      </ButtonGroup>
    </React.Fragment>
  )
}

export default ButtonDemo;
