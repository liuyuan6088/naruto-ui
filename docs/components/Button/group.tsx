import * as React from 'react';
import { Button } from 'components';
import './index.less';

const ButtonGroup = Button.Group;

const Group: React.FC = () => {

  return (
    <React.Fragment>

      <p id="screens">按钮组合</p>
      <div className='button-box'>
        <ButtonGroup>
          {['left', 'mid', 'right'].map((e) => (
            <Button
              key={e}
              type='primary'
            >
              {e}
            </Button>
          ))}
        </ButtonGroup>

        <ButtonGroup>
          {['lg', 'nm', 'sm'].map((e) => (
            <Button
              key={e}
            >
              {e}
            </Button>
          ))}
        </ButtonGroup>
      </div>

      
    </React.Fragment>
  )
}

export default Group;
