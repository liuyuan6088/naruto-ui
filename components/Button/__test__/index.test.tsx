import * as React from 'react'
import { render } from 'react-testing-library'
import Button from '../index'

describe('render button', () => {
  it('render default', () => {
    const wrapper = render(<Button />)
    expect(wrapper.container).toMatchSnapshot()
  })
  it('render button with children', () => {
    const childrenBtn = render(<Button>чбошод</Button>)
    expect(childrenBtn.container).toMatchSnapshot()
  })
})
