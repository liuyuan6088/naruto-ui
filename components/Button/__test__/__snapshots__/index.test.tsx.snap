// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render button render button with type-icon 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "naruto-btn naruto-btn-default",
      "type": "button",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "naruto-icon naruto-icon-thunderbolt naruto-icon-spin naruto-btn-icon",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "href": "#icon-thunderbolt",
            },
            "children": Array [],
            "name": "use",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "href": "http://www.w3.org/1999/xlink",
            },
            "x-attribsPrefix": Object {
              "href": "xlink",
            },
          },
        ],
        "name": "svg",
        "namespace": "http://www.w3.org/2000/svg",
        "next": Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "霹雳",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "霹雳",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "naruto-icon naruto-icon-thunderbolt naruto-icon-spin naruto-btn-icon",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "href": "#icon-thunderbolt",
              },
              "children": Array [],
              "name": "use",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "href": "http://www.w3.org/1999/xlink",
              },
              "x-attribsPrefix": Object {
                "href": "xlink",
              },
            },
          ],
          "name": "svg",
          "namespace": "http://www.w3.org/2000/svg",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`render button render default 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "naruto-btn naruto-btn-default",
      "type": "button",
    },
    "children": Array [
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "Follow",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`render button render loading button 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "naruto-btn naruto-btn-loading naruto-btn-default",
      "type": "button",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "naruto-icon naruto-icon-reload naruto-icon-spin naruto-btn-icon",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "href": "#icon-reload",
            },
            "children": Array [],
            "name": "use",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "href": "http://www.w3.org/1999/xlink",
            },
            "x-attribsPrefix": Object {
              "href": "xlink",
            },
          },
        ],
        "name": "svg",
        "namespace": "http://www.w3.org/2000/svg",
        "next": Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "loading",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "loading",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "naruto-icon naruto-icon-reload naruto-icon-spin naruto-btn-icon",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "href": "#icon-reload",
              },
              "children": Array [],
              "name": "use",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "href": "http://www.w3.org/1999/xlink",
              },
              "x-attribsPrefix": Object {
                "href": "xlink",
              },
            },
          ],
          "name": "svg",
          "namespace": "http://www.w3.org/2000/svg",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
